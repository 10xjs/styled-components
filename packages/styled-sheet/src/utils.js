// @flow

import { SC_ATTR, SC_ATTR_VERSION, SC_VERSION, SC_ACTIVE, NONCE } from './constants';

export const setStyledAttributes = (tag: HTMLElement): void => {
  tag.setAttribute(SC_ATTR, SC_ACTIVE);
  tag.setAttribute(SC_ATTR_VERSION, SC_VERSION);
  tag.setAttribute('nonce', NONCE);
};

export const makeStyleTag = (target?: HTMLElement): HTMLStyleElement => {
  // $FlowFixMe
  const head = (document.head: HTMLElement);
  const element = document.createElement('style');
  setStyledAttributes(element);
  (target || head).appendChild(element);
  return element;
};

export const getSheet = (tag: HTMLStyleElement): CSSStyleSheet => {
  if (tag.sheet) {
    // $FlowFixMe
    return (tag.sheet: CSSStyleSheet);
  }

  /* Firefox quirk requires us to step through all stylesheets to find one owned by the given tag */
  const { styleSheets } = document;
  for (let i = 0, l = styleSheets.length; i < l; i++) {
    const sheet = styleSheets[i];
    if (sheet.ownerNode === tag) {
      // $FlowFixMe
      return (sheet: CSSStyleSheet);
    }
  }

  throw new TypeError(`CSSStyleSheet could not be found on HTMLStyleElement`);
};

export const makeCssMarker = (name: string, group: number, keys: string[]) => `/*sc-${group}:${name}:${keys.join(',')}*/`;

export const cssMarkerRe = new RegExp(/\/\*sc-(\d+):([^:]+):([^*]*)\*\//, 'g');

export const wrapInStyleTag = (css: string): string => {
  const opening = `<style ${SC_ATTR} ${SC_ATTR_VERSION}="${SC_VERSION}">\n`;
  const closing = `</style>`;
  return `${opening}${css}${closing}`;
};
