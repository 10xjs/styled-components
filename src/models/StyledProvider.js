// @flow

import React, { Component, createContext } from 'react';

const StyledContext = createContext();

export const { Consumer: StyledConsumer } = StyledContext;

export default class StyledProvider extends Component<*> {
  static defaultProps = {
    scope: '',
  };

  render() {
    const { scope, ...restProps } = this.props;
    return <StyledContext.Provider value={{ scope }} {...restProps} />;
  }
}
