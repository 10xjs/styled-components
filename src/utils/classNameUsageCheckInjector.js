// @flow
import getComponentName from './getComponentName';

export default (target: Object) => {
  let elementClassName = '';
  let elementRef = null;
  let node: Element | null = null;

  const targetCDM = target.componentDidMount;

  // eslint-disable-next-line no-param-reassign
  target.componentDidMount = function componentDidMount() {
    if (typeof targetCDM === 'function') {
      targetCDM.call(this);
    }

    const classNames = elementClassName
      .replace(/ +/g, ' ')
      .trim()
      .split(' ');

    // eslint-disable-next-line react/no-find-dom-node
    const selector = classNames.map(s => `.${s}`).join('');

    if (
      node &&
      node.nodeType === 1 &&
      !classNames.every(
        className => node && node.classList && node.classList.contains(className)
      ) &&
      !node.querySelector(selector)
    ) {
      console.warn(
        `It looks like you've wrapped styled() around your React component (${getComponentName(
          this.props.forwardedClass.target
        )}), but the className prop is not being passed down to a child. No styles will be rendered unless className is composed within your React component.`
      );
    }
  };

  const prevRenderInner = target.renderInner;

  // eslint-disable-next-line no-param-reassign
  target.renderInner = function renderInner(...args) {
    elementRef = target.props.forwardedRef || target.props.ref;

    // https://github.com/facebook/react/issues/13029
    // eslint-disable-next-line no-param-reassign
    target.internalRef = el => {
      node = el;

      if (elementRef) {
        if (typeof elementRef === 'function') {
          elementRef(el);
        } else if (typeof elementRef === 'object') {
          elementRef.current = el;
        }
      }
    };

    const element = prevRenderInner.apply(this, args);

    elementClassName = element.props.className;

    return element;
  };
};
