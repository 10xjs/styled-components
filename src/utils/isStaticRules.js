// @flow
import isFunction from './isFunction';
import isStyledComponent from './isStyledComponent';
import type { RuleSet } from '../types';

export default function isStaticRules(rules: RuleSet, attrsSet?: Array<Object | Function>): boolean {
  for (let i = 0; i < rules.length; i += 1) {
    const rule = rules[i];

    // recursive case
    if (Array.isArray(rule) && !isStaticRules(rule)) {
      return false;
    } else if (isFunction(rule) && !isStyledComponent(rule)) {
      // functions are allowed to be static if they're just being
      // used to get the classname of a nested styled component
      return false;
    }
  }

  if (Array.isArray(attrsSet) && attrsSet.length > 0) {
    for (let i = 0; i < attrsSet.length; i += 1) {
      const attrs = attrsSet[i];
      if (isFunction(attrs)) {
        return false;
      }
    }
    return false;
  }

  return true;
}
