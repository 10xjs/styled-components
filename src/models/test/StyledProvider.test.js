// @flow
/* eslint-disable react/no-multi-comp */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import StyledProvider from '../StyledProvider';
import withTheme from '../../hoc/withTheme';
import { resetStyled } from '../../test/utils';

let styled;

describe('StyledProvider', () => {
  beforeEach(() => {
    styled = resetStyled();
  });

  it('should not throw an error when no children are passed', () => {
    TestRenderer.create(<StyledProvider scope="" />);
  });

  it("should accept a scope prop that's a string", () => {
    TestRenderer.create(<StyledProvider scope=".scope" />);
  });

  it('should render its child', () => {
    const child = <p>Child!</p>;
    const wrapper = TestRenderer.create(<StyledProvider scope=".scope">{child}</StyledProvider>);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should scope styles to the prefix specified in the scope prop', () => {
    const MyDiv = styled.div``;
    const child = <MyDiv>Child!</MyDiv>;
    const wrapper = TestRenderer.create(<StyledProvider scope=".scope">{child}</StyledProvider>);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
