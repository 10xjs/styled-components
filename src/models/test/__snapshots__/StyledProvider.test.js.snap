// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyledProvider should render its child 1`] = `
<p>
  Child!
</p>
`;

exports[`StyledProvider should scope styles to the prefix specified in the scope prop 1`] = `
<div
  className="sc-a b"
>
  Child!
</div>
`;
